stdlib/public/runtime/CrashHandlerLinux.cpp

```diff
@@ -0,0 +1,770 @@
+}
+
+bool
+seen_thread(pid_t tid) {
+  uint64_t next = __atomic_load_n(&crashInfo.thread_list, __ATOMIC_ACQUIRE);
+  while (next) {
+    struct thread *pthread = (struct thread *)next;
+    if (pthread->tid == tid)
+      return true;
+    next = pthread->next;
+  }
+  return false;
+}
+
+void
+pause_thread(int signum __attribute__((unused)),
+             siginfo_t *pinfo  __attribute__((unused)),
+             void *uctx)
+{
+  int old_err = errno;
+  struct thread self = { 0, (int64_t)gettid(), (uint64_t)uctx };
+
+  add_thread(&self);
+
+  notify_paused();
+
+  take_thread_lock();
+  release_thread_lock();
+
+  errno = old_err;
+}
+
+struct linux_dirent64 {
+  ino64_t        d_ino;
+  off64_t        d_off;
+  unsigned short d_reclen;
+  unsigned char  d_type;
+  char           d_name[256];
+};
+
+int
+getdents(int fd, void *buf, size_t bufsiz)
+{
+  return syscall(SYS_getdents64, fd, buf, bufsiz);
+}
+
+/* Stop all other threads in this process; we do this by establishing a
+   signal handler for SIGUSR1, then iterating through the threads sending
+   SIGUSR1.
+
+   Finding the other threads is a pain, because Linux has no actual API
+   for that; instead, you have to read /proc.  Unfortunately, opendir()
+   and readdir() are not async signal safe, so we get to do this with
+   the getdents system call instead.
+
+   The SIGUSR1 signals also serve to build the thread list. */
+void
+suspend_other_threads(struct thread *self)
+{
+  struct sigaction sa, sa_old;
+
+  // Take the lock
+  take_thread_lock();
+
+  // Start the thread list with this thread
+  reset_threads(self);
+
+  // Swap out the SIGPROF signal handler first
+  sigfillset(&sa.sa_mask);
+  sa.sa_flags = SA_NODEFER;
+  sa.sa_handler = NULL;
+  sa.sa_sigaction = pause_thread;
+
+  sigaction(SIGPROF, &sa, &sa_old);
+
+  /* Now scan /proc/self/task to get the tids of the threads in this
+     process.  We need to ignore our own thread. */
+  int fd = open("/proc/self/task",
+                O_RDONLY|O_NDELAY|O_DIRECTORY|O_LARGEFILE|O_CLOEXEC);
+  int our_pid = getpid();
+  char buffer[4096];
+  size_t offset = 0;
+  size_t count = 0;
+
+  uint32_t thread_count = 0;
+  uint32_t old_thread_count;
+
+  do {
+    old_thread_count = thread_count;
+    lseek(fd, 0, SEEK_SET);
+
+    for (;;) {
+      if (offset >= count) {
+        ssize_t bytes = getdents(fd, buffer, sizeof(buffer));
+        if (bytes <= 0)
+          break;
+        count = (size_t)bytes;
+        offset = 0;
+      }
+
+      struct linux_dirent64 *dp = (struct linux_dirent64 *)&buffer[offset];
+      offset += dp->d_reclen;
+
+      if (strcmp(dp->d_name, ".") == 0
+          || strcmp(dp->d_name, "..") == 0)
+        continue;
+
+      int tid = atoi(dp->d_name);
+
+      if ((int64_t)tid != self->tid && !seen_thread(tid)) {
+        tgkill(our_pid, tid, SIGPROF);
+        ++thread_count;
+      }
+    }
+
+    // Wait up to 5 seconds for the threads to pause
```

Comment:
I also need to update the comment here, I think. It still says `SIGUSR1` but it's now using `SIGPROF` because using the former breaks when `libdispatch` is in use.

Reply:
(Yes, this is truly horrid. I don't see any other possibility though. We can definitely get threads into an interruptible wait, and in that case they won't get the signal until that is over, and we don't want to wait indefinitely.)